File: source_code/project4_spring/main/src/main/java/com/service/main/entity/Booking.java
package com.service.main.entity;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import jakarta.persistence.*;
import jakarta.persistence.criteria.CriteriaBuilder;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.Date;
import java.util.List;

@Entity
@Data
@AllArgsConstructor
@NoArgsConstructor
public class Booking {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY )
    private Integer id;

    private Date checkInDay;

    private Date checkOutDay;

    private int totalPerson;

    private int children;

    private int adult;

    private String bookingType;

    private boolean isSelfCheckIn;

    private String selfCheckInType;

    @Column(columnDefinition = "TEXT")
    private String selfCheckInInstruction;

    private double amount;

    private double websiteFee;

    private double hostFee;

    private LocalDateTime createdAt =  LocalDateTime.now();

    private LocalDateTime updatedAt =  LocalDateTime.now();

    private String status;

    private String bookingCode;

    @ManyToOne
    @JoinColumn(name = "propertyId")
    @JsonBackReference
    private Property property;

    @ManyToOne
    @JoinColumn(name = "refundPolicyId")
    @JsonBackReference
    private RefundPolicy refundPolicy;

    @OneToOne
    @JoinColumn(name = "hostReviewId")
    @JsonBackReference
    private Review hostReview;

    @OneToOne
    @JoinColumn(name = "userReviewId")
    @JsonBackReference
    private Review userReview;

    @ManyToOne
    @JoinColumn(name = "hostId")
    @JsonBackReference
    private User host;

    @ManyToOne
    @JoinColumn(name = "customerId")
    @JsonBackReference
    private User customer;

    @OneToMany(mappedBy = "booking", fetch = FetchType.LAZY, cascade = CascadeType.ALL)
    @JsonManagedReference
    private List<Transaction> transactions;

    @OneToMany(mappedBy = "booking", fetch = FetchType.LAZY, cascade = CascadeType.ALL)
    @JsonManagedReference
    private List<BookDateDetail> bookDateDetails;

    @ManyToOne
    @JoinColumn(name = "discountId")
    @JsonBackReference
    private Discount discount;


}

File: source_code/project4_flutter/lib/features/property_detail/models/booking.dart
import 'package:json_annotation/json_annotation.dart';
part 'booking.g.dart';

@JsonSerializable()
class Booking {
  final int? id;
  final DateTime? checkInDay;
  final DateTime? checkOutDay;
  final int? totalPerson;
  final int? children;
  final int? adult;
  final double? amount;
  final String? bookingType;

  final bool? selfCheckIn;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final String? status;
  final String? bookingCode;
  final int? websiteFee;
  final int? hostFee;
  Booking({
    this.id,
    this.checkInDay,
    this.checkOutDay,
    this.totalPerson,
    this.children,
    this.adult,
    this.bookingType,
    this.selfCheckIn,
    this.amount,
    this.createdAt,
    this.updatedAt,
    this.status,
    this.bookingCode,
    this.websiteFee,
    this.hostFee,
  });
  factory Booking.fromJson(Map<String, dynamic> json) =>
      _$BookingFromJson(json);
  Map<String, dynamic> toJson() => _$BookingToJson(this);
}
